### ISSUE TITLE
Feature Request: Support field numbers as keys for json_format.MessageToJson and MessageToDict

### ISSUE THREAD

**Issue by kunalsinghal** [protocolbuffers/protobuf#15146]
**What language does this apply to?**

Python protobuf generated code

**Describe the problem you are trying to solve.**

I want to be able to serialize protobuf messages to JSON but use field numbers as keys instead of field name. Field names don't support field renaming as a backward-compatible change if you're using JSON serialization. 

**Describe the solution you'd like**

```python
def MessageToJson(
    message,
    including_default_value_fields=False,
    preserving_proto_field_name=False,
    indent=2,
    sort_keys=False,
    use_integers_for_enums=False,
    descriptor_pool=None,
    float_precision=None,
    ensure_ascii=True,
    use_field_numbers_for_keys=false):
```

Add a new kwarg to `MessageToJson` and `MessageToDict` to use field numbers instead of names.

**Describe alternatives you've considered**

Been using field name based JSONs so far. It's more readable serialization but not backward compatible with renames.

**Additional context**

n/a

**Comment by esrauchg** [comment]
I agree that it would have been better if the JSON format was spec'd to have parsers accept field numbers as keys, and serializers to have an option to emit that way.

The reason why it wasn't was the most common expected usecase of the JSON format was expected to be used where only one side was a Protobuf library (either only the client or only the server) and the other side not, and that field names would be realistically needed compared for ergonomics. In practice many people do use a Protobuf library on both ends with the ProtoJSON format.

Unfortunately in 2026+ the difficulty of this is rolling out such a change; we highly value the property that the ecosystem has a lot of compatible implementations which match spec, and are very wary about things which would fragment the ecosystem such that implementation X and implementation Y can't interoperate.

I think even though it makes sense, we're unlikely to change the JSON format spec in 2026 for this. If we do an overhaul where we change many other things in the format to match more recent JSON idioms that have emerged since ProtoJSON was designed in 2014 then I think this might make sense as a 'free rider'.
